File: java/playwright/src/test/java/dev/selenium/request/ModifyHeaderTest.java
package dev.selenium.request;

import com.microsoft.playwright.Route;
import dev.selenium.TestBase;
import java.util.Map;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class ModifyHeaderTest extends TestBase {

  @Test
  public void addHeader() {
    page.route(
        "https://httpbin.org/**",
        route -> {
          Map<String, String> headers = route.request().headers();
          headers.put("X-Test", "true");
          route.resume(new Route.ResumeOptions().setHeaders(headers));
        });

    page.navigate("https://httpbin.org/headers");

    String bodyText = page.innerText("body").replaceAll("\\s+", "");
    Assertions.assertTrue(bodyText.contains("\"X-Test\":\"true\""));
  }

  @Test
  public void removeHeader() {
    page.route(
        "https://httpbin.org/**",
        route -> {
          Map<String, String> headers = route.request().headers();
          headers.remove("upgrade-insecure-requests");
          route.resume(new Route.ResumeOptions().setHeaders(headers));
        });

    page.navigate("https://httpbin.org/headers");

    Assertions.assertFalse(page.innerText("body").contains("Upgrade-Insecure-Requests"));
  }
}

File: dotnet/Playwright/Request/ModifyHeaderTest.cs
using Microsoft.Playwright;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Playwright.Request;

[TestClass]
public class ModifyHeaderTest : TestBase
{
    [TestMethod]
    public async Task AddHeader()
    {
        await Page!.RouteAsync("https://httpbin.org/**", async route =>
        {
            var headers = new Dictionary<string, string>(await route.Request.AllHeadersAsync());
            headers["X-Test"] = "true";
            await route.ContinueAsync(new RouteContinueOptions { Headers = headers });
        });

        await Page.GotoAsync("https://httpbin.org/headers");

        var bodyText = await Page.InnerTextAsync("body");
        Assert.IsTrue(bodyText.Contains("\"X-Test\": \"true\""));
    }

    [TestMethod]
    public async Task RemoveHeader()
    {
        await Page!.RouteAsync("https://httpbin.org/**", async route =>
        {
            var headers = new Dictionary<string, string>(await route.Request.AllHeadersAsync());
            headers.Remove("upgrade-insecure-requests");
            await route.ContinueAsync(new RouteContinueOptions { Headers = headers });
        });

        await Page.GotoAsync("https://httpbin.org/headers");

        var bodyText = await Page.InnerTextAsync("body");
        Assert.IsFalse(bodyText.Contains("Upgrade-Insecure-Requests"));
    }
}